src/tools/r2rdump/EHInfo.cs

```diff
@@ -0,0 +1,237 @@
+        /// </summary>
+        public uint HandlerEnd;
+
+        /// <summary>
+        /// For type-based exception handlers, this is the type token.
+        /// For filter-based exception handlers, this is the filter offset.
+        /// </summary>
+        public uint ClassTokenOrFilterOffset;
+
+        /// <summary>
+        /// Read the EH clause from a given file offset in the PE image.
+        /// </summary>
+        /// <param name="image">R2R PE executable image</param>
+        /// <param name="offset">Offset of the EH clause in the image</param>
+        public EHClause(byte[] image, int offset)
+        {
+            Flags = (CorExceptionFlag)BitConverter.ToUInt32(image, offset + 0 * sizeof(uint));
+            TryOffset = BitConverter.ToUInt32(image, offset + 1 * sizeof(uint));
+            TryEnd = BitConverter.ToUInt32(image, offset + 2 * sizeof(uint));
+            HandlerOffset = BitConverter.ToUInt32(image, offset + 3 * sizeof(uint));
+            HandlerEnd = BitConverter.ToUInt32(image, offset + 4 * sizeof(uint));
+            ClassTokenOrFilterOffset = BitConverter.ToUInt32(image, offset + 5 * sizeof(uint));
+        }
+
+        /// <summary>
+        /// Emit a textual representation of the EH info to a given string builder.
+        /// </summary>
+        /// <param name="stringBuilder">Output builder for the textual representation</param>
+        /// <param name="methodRva">Starting RVA of the runtime function is used to display the try / handler info as RVA intervals</param>
+        public void WriteTo(StringBuilder stringBuilder, int methodRva)
+        {
+            stringBuilder.Append($@"Flags {(uint)Flags:X2} ");
+            stringBuilder.Append($@"TryOff {TryOffset:X4} (RVA {(TryOffset + methodRva):X4}) ");
+            stringBuilder.Append($@"TryEnd {TryEnd:X4} (RVA {(TryEnd + methodRva):X4}) ");
+            stringBuilder.Append($@"HndOff {HandlerOffset:X4} (RVA {(HandlerOffset + methodRva):X4}) ");
+            stringBuilder.Append($@"HndEnd {HandlerEnd:X4} (RVA {(HandlerEnd + methodRva):X4}) ");
+            stringBuilder.Append($@"ClsFlt {ClassTokenOrFilterOffset:X4}");
+
+            if ((Flags & CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_FILTER) != CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_NONE)
+            {
+                stringBuilder.Append($@" (RVA {(ClassTokenOrFilterOffset + methodRva):X4})");
+
+                stringBuilder.Append(" FILTER");
+            }
+            if ((Flags & CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_FINALLY) != CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_NONE)
+            {
+                stringBuilder.Append(" FINALLY");
+            }
+            if ((Flags & CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_FAULT) != CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_NONE)
+            {
+                stringBuilder.Append(" FAULT");
+            }
+            if ((Flags & CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_DUPLICATED) != CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_NONE)
```

Comment:
I think these checks are slightly wrong. The definition is quite weird, since only one of NONE, FILTER, FINALLY, and FAULT can be true at a time. Thus, you can do a bit check. Or, you can simply mask 0x7 and direct compare. However, DUPLICATED can be set at the same time as one of the others (e.g., DUPLICATED as well as FILTER). So just do a bit check.

Seems like for NONE you should explicitly say it's a class handler, and maybe (if possible) construct the class name corresponding to the ClassToken.

Reply:
Thanks @BruceForstall, that's a very good suggestion, will fix!

Reply:
Fixed in 3rd commit.